File: 01297. Maximum Number of Occurrences of a Substring/solution.go
/**
 * Problem: 1297. Maximum Number of Occurrences of a Substring
 *
 * Difficulty: Medium
 *
 * Language: Golang
 *
 * Performance: Runtime - 9 ms (Beats 100%)
 */

func maxFreq(s string, maxLetters int, minSize int, maxSize int) int {
	// Check if the input string is empty or any parameter is invalid
	if s == "" || maxLetters <= 0 || minSize <= 0 || maxSize <= 0 {
		return 0
	}

	// If the string length is less than the minimum size, return 0
	if len(s) < minSize {
		return 0
	}

	// Initialize a map to store substring frequencies
	freqs := map[string]int{}

	// Initialize an array to count character occurrences
	var cnt [26]int

	// Variable to track the number of distinct characters in the window
	distincts := 0

	// Populate the initial window of size `minSize`
	for i := 0; i < minSize; i++ {
		idx := s[i] - 'a'
		cnt[idx]++

		// Increment distinct character count if this is the first occurrence
		if cnt[idx] == 1 {
			distincts++
		}
	}

	// Add the initial substring to the frequency map if it meets the criteria
	if distincts <= maxLetters {
		subStr := s[0:minSize]
		freqs[subStr] += 1
	}

	// Slide the window across the string
	for r := minSize; r < len(s); r++ {
		// Remove the leftmost character from the window
		left := r - minSize
		lIdx := s[left] - 'a'
		cnt[lIdx]--

		// Decrease distinct character count if the character count becomes zero
		if cnt[lIdx] == 0 {
			distincts--
		}

		// Add the new rightmost character to the window
		rIdx := s[r] - 'a'
		cnt[rIdx] += 1

		// Increment distinct character count if this is the first occurrence
		if cnt[rIdx] == 1 {
			distincts++
		}

		// Add the current substring to the frequency map if it meets the criteria
		if distincts <= maxLetters {
			subStr := s[left+1 : r+1]
			freqs[subStr] += 1
		}
	}

	// Initialize the result variable to store the maximum frequency
	ans := 0

	// Iterate through the frequency map to find the maximum frequency
	for _, v := range freqs {
		if ans < v {
			ans = v
		}
	}

	// Return the maximum frequency of valid substrings
	return ans
}

File: 01297. Maximum Number of Occurrences of a Substring/solution.js
/**
 * Problem: 1297. Maximum Number of Occurrences of a Substring
 *
 * Difficulty: Medium
 *
 * Language: JavaScript
 *
 * Performance: Runtime - 25 ms (Beats 100%)
 */

/**
 * Finds the max frequency of substrings meeting the criteria
 *
 * @param {string} s - Input string
 * @param {number} maxLetters - Max unique letters in substring
 * @param {number} minSize - Min substring length
 * @param {number} maxSize - Max substring length
 *
 * @returns {number} - Max frequency of valid substrings
 */
const maxFreq = (s, maxLetters, minSize, maxSize) => {
  // Initialize an array to count occurrences of each letter (26 letters in alphabet)
  const cnt = new Array(26).fill(0)

  // Initialize variables to track the number of unique letters and the max frequency
  let uniqueLetterCount = 0,
    maxFrequency = 0

  // Create a map to store the frequency of substrings
  const substringFrequency = new Map()

  // Get the ASCII code of 'a' for character calculations
  const aCharCode = 'a'.charCodeAt(0)

  // Function to add a character to the count and update unique letter count
  const addCharacter = (character) => {
    const index = character.charCodeAt(0) - aCharCode

    // Increment the count for the character and check if it's a new unique letter
    if (cnt[index]++ === 0) uniqueLetterCount++
  }

  // Function to remove a character from the count and update unique letter count
  const removeCharacter = (character) => {
    const index = character.charCodeAt(0) - aCharCode

    // Decrement the count for the character and check if it is no longer unique
    if (--cnt[index] === 0) uniqueLetterCount--
  }

  // Iterate through the input string
  for (let i = 0; i < s.length; i++) {
    // Add the current character to the count
    addCharacter(s[i])

    // Remove the character that is out of the current window (if applicable)
    if (i >= minSize) removeCharacter(s[i - minSize])

    // Check if the current window meets the criteria for a valid substring
    if (i + 1 >= minSize && uniqueLetterCount <= maxLetters) {
      // Extract the current substring
      const substring = s.slice(i - minSize + 1, i + 1)

      // Update the frequency of the substring in the map
      const frequency = (substringFrequency.get(substring) || 0) + 1

      // Set the updated frequency back in the map
      substringFrequency.set(substring, frequency)

      // Update the maximum frequency if the current substring frequency is higher
      if (frequency > maxFrequency) maxFrequency = frequency
    }
  }

  // Return the maximum frequency of valid substrings
  return maxFrequency
}
